### ISSUE TITLE
RedirectUriReceiverActivity not always being triggered with chrome 53

### ISSUE THREAD

**Issue by mattinger** [openid/AppAuth-Android#114]
I realize this is likely a chrome bug, but when using App Auth with chrome 53 (i've tried this on both a nexus 5x and a nexus6 with the same result).  After authenticating via a pass though (ie. with prompt=none, or an existing browser cookie with no prompt value), the browser window will close, but RedirectUriReceiverActivity is not triggering.

However, if i roll back to chrome 51, everything works fine.

Also, using firefox or the system browser doesn't seem to be suffering from the same issue.

I'm seeing the following in the system logs:

```
10-13 09:11:49.850 28569-28569/? W/Thread-3: type=1400 audit(0.0:244): avc: denied { ioctl } for path="socket:[466550]" dev="sockfs" ino=466550 ioctlcmd=8b1b scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:r:untrusted_app:s0:c512,c768 tclass=udp_socket permissive=0
10-13 09:11:49.859 28508-28569/? D/cr_SafeBrowsingApi: [SafeBrowsingApiHandlerInternal.java:136] Starting request.
10-13 09:11:49.869 28508-28569/? D/cr_SafeBrowsingApi: [SafeBrowsingApiHandlerInternal.java:145] Done starting request
10-13 09:11:49.870 28508-28508/? D/cr_SafeBrowsingApi: [SafeBrowsingApiHandlerInternal.java:116] Done calling callback
10-13 09:11:49.870 28508-28569/? D/cr_SafeBrowsingApi: [SafeBrowsingApiBridge.java:62] Done starting request
10-13 09:11:51.304 28569-28569/? W/Thread-3: type=1400 audit(0.0:245): avc: denied { ioctl } for path="socket:[468157]" dev="sockfs" ino=468157 ioctlcmd=8b1b scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:r:untrusted_app:s0:c512,c768 tclass=udp_socket permissive=0
10-13 09:11:51.317 871-2934/? I/ActivityManager: START u0 {act=android.intent.action.VIEW cat=[android.intent.category.BROWSABLE] dat=myapp://auth?code=0.ac.qddqkI&state=s46UZ53hhjoB4HsVZ2YLGA flg=0x10000000 cmp=com.myapp/net.openid.appauth.RedirectUriReceiverActivity (has extras)} from uid 10041 on display 0
10-13 09:11:51.337 871-3025/? I/ActivityManager: moveTaskToBack: TaskRecord{f5b15c7 #308 A=com.myapp U=0 StackId=1 sz=3}
10-13 09:11:51.359 28508-28508/? W/cr_ScreenOrientation: Removing an inexistent observer!
10-13 09:11:51.379 28508-28508/? D/cr_GCore: [LifecycleHook.java:73] onApplicationStateChange
10-13 09:11:51.380 28508-28611/? E/ConnectivityManager.CallbackHandler: callback not found for RELEASED message
10-13 09:11:51.385 399-399/? E/lowmemorykiller: Error writing /proc/29481/oom_score_adj; errno=22
10-13 09:11:51.400 871-3534/? I/ActivityManager: Process com.android.chrome:sandboxed_process1 (pid 29481) has died
10-13 09:11:51.400 871-3534/? D/ActivityManager: cleanUpApplicationRecord -- 29481
10-13 09:11:51.400 871-3534/? W/ActivityManager: Scheduling restart of crashed service com.android.chrome/org.chromium.content.app.SandboxedProcessService1 in 1000ms
10-13 09:11:51.420 28508-28508/? W/cr_ChildProcessConnect: onServiceDisconnected (crash or killed by oom): pid=29481
10-13 09:11:51.423 28508-28508/? D/cr_ChildProcLauncher: [ChildProcessLauncher.java:878] stopping child connection: pid=29481
10-13 09:11:51.424 28508-28508/? D/cr_ChildProcLauncher: [ChildProcessLauncher.java:119] Allocator freed a connection, sandbox: true, slot: 1
10-13 09:11:51.722 28508-28508/? D/cr_GCore: [LifecycleHook.java:73] onApplicationStateChange
10-13 09:11:51.723 28508-28508/? D/cr_GCore: [LifecycleHook.java:79] Application visibilty changed to false. Updating state of 1 client(s).
10-13 09:11:51.742 28508-28508/? D/cr_GCore: [LifecycleHook.java:73] onApplicationStateChange
10-13 09:11:51.886 871-891/? I/WindowManager: Destroying surface Surface(name=SurfaceView - com.android.chrome/org.chromium.chrome.browser.customtabs.CustomTabActivity) called by com.android.server.wm.WindowStateAnimator.destroySurface:2014 com.android.server.wm.WindowStateAnimator.destroySurfaceLocked:881 com.android.server.wm.WindowState.removeLocked:1449 com.android.server.wm.WindowManagerService.removeWindowInnerLocked:2478 com.android.server.wm.AppWindowToken.destroySurfaces:365 com.android.server.wm.WindowStateAnimator.finishExit:565 com.android.server.wm.WindowStateAnimator.finishExit:516 com.android.server.wm.WindowStateAnimator.stepAnimationLocked:491 
10-13 09:11:51.898 871-891/? I/WindowManager: Destroying surface Surface(name=com.android.chrome/org.chromium.chrome.browser.customtabs.CustomTabActivity) called by com.android.server.wm.WindowStateAnimator.destroySurface:2014 com.android.server.wm.WindowStateAnimator.destroySurfaceLocked:881 com.android.server.wm.WindowState.destroyOrSaveSurface:2073 com.android.server.wm.AppWindowToken.destroySurfaces:363 com.android.server.wm.WindowStateAnimator.finishExit:565 com.android.server.wm.WindowStateAnimator.stepAnimationLocked:491 com.android.server.wm.WindowAnimator.updateWindowsLocked:303 com.android.server.wm.WindowAnimator.animateLocked:704 
10-13 09:11:51.901 401-401/? W/SurfaceFlinger: couldn't log to binary event log: overflow.
10-13 09:11:56.812 28508-28508/? D/cr_GCore: [GoogleApiClientHelper.java:185] Disconnect delay expired.
```

**Comment by iainmcgin** [comment]
I have reached out to the maintainers of Chrome Custom Tabs to investigate this further, thanks for reporting the issue.

**Comment by iainmcgin** [comment]
The Chrome team got back to me: they believe this issue has already been fixed with a maintenance release of Chrome 53, so try forcing an update. They also suggested that you try testing Chrome 54 beta, as this definitely has the fix they are talking about. 

If neither of those work, let me know, and I'll do some more digging.

**Comment by ccarrasc** [comment]
I was seeing a similar issue as well. Just tried chrome beta (54) and it appears to work just fine. Thanks!

**Comment by mattinger** [comment]
Chrome 54 beta solves it for me as well.  Thanks

**Comment by iainmcgin** [comment]
For posterity, the specific version of Chrome from which this should be fixed is `53.0.2785.124`.

**Comment by iainmcgin** [comment]
Browser filtering is now available in HEAD (and will be included in the 0.4.0 release), allowing you to blacklist versions of browsers and custom tab usage that misbehave.

**Comment by ALBARAKPP** [comment]
> I realize this is likely a chrome bug, but when using App Auth with chrome 53 (i've tried this on both a nexus 5x and a nexus6 with the same result).  After authenticating via a pass though (ie. with prompt=none, or an existing browser cookie with no prompt value), the browser window will close, but RedirectUriReceiverActivity is not triggering.
> 
> However, if i roll back to chrome 51, everything works fine.
> 
> Also, using firefox or the system browser doesn't seem to be suffering from the same issue.
> 
> I'm seeing the following in the system logs:
> 
> ```
> 10-13 09:11:49.850 28569-28569/? W/Thread-3: type=1400 audit(0.0:244): avc: denied { ioctl } for path="socket:[466550]" dev="sockfs" ino=466550 ioctlcmd=8b1b scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:r:untrusted_app:s0:c512,c768 tclass=udp_socket permissive=0
> 10-13 09:11:49.859 28508-28569/? D/cr_SafeBrowsingApi: [SafeBrowsingApiHandlerInternal.java:136] Starting request.
> 10-13 09:11:49.869 28508-28569/? D/cr_SafeBrowsingApi: [SafeBrowsingApiHandlerInternal.java:145] Done starting request
> 10-13 09:11:49.870 28508-28508/? D/cr_SafeBrowsingApi: [SafeBrowsingApiHandlerInternal.java:116] Done calling callback
> 10-13 09:11:49.870 28508-28569/? D/cr_SafeBrowsingApi: [SafeBrowsingApiBridge.java:62] Done starting request
> 10-13 09:11:51.304 28569-28569/? W/Thread-3: type=1400 audit(0.0:245): avc: denied { ioctl } for path="socket:[468157]" dev="sockfs" ino=468157 ioctlcmd=8b1b scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:r:untrusted_app:s0:c512,c768 tclass=udp_socket permissive=0
> 10-13 09:11:51.317 871-2934/? I/ActivityManager: START u0 {act=android.intent.action.VIEW cat=[android.intent.category.BROWSABLE] dat=myapp://auth?code=0.ac.qddqkI&state=s46UZ53hhjoB4HsVZ2YLGA flg=0x10000000 cmp=com.myapp/net.openid.appauth.RedirectUriReceiverActivity (has extras)} from uid 10041 on display 0
> 10-13 09:11:51.337 871-3025/? I/ActivityManager: moveTaskToBack: TaskRecord{f5b15c7 #308 A=com.myapp U=0 StackId=1 sz=3}
> 10-13 09:11:51.359 28508-28508/? W/cr_ScreenOrientation: Removing an inexistent observer!
> 10-13 09:11:51.379 28508-28508/? D/cr_GCore: [LifecycleHook.java:73] onApplicationStateChange
> 10-13 09:11:51.380 28508-28611/? E/ConnectivityManager.CallbackHandler: callback not found for RELEASED message
> 10-13 09:11:51.385 399-399/? E/lowmemorykiller: Error writing /proc/29481/oom_score_adj; errno=22
> 10-13 09:11:51.400 871-3534/? I/ActivityManager: Process com.android.chrome:sandboxed_process1 (pid 29481) has died
> 10-13 09:11:51.400 871-3534/? D/ActivityManager: cleanUpApplicationRecord -- 29481
> 10-13 09:11:51.400 871-3534/? W/ActivityManager: Scheduling restart of crashed service com.android.chrome/org.chromium.content.app.SandboxedProcessService1 in 1000ms
> 10-13 09:11:51.420 28508-28508/? W/cr_ChildProcessConnect: onServiceDisconnected (crash or killed by oom): pid=29481
> 10-13 09:11:51.423 28508-28508/? D/cr_ChildProcLauncher: [ChildProcessLauncher.java:878] stopping child connection: pid=29481
> 10-13 09:11:51.424 28508-28508/? D/cr_ChildProcLauncher: [ChildProcessLauncher.java:119] Allocator freed a connection, sandbox: true, slot: 1
> 10-13 09:11:51.722 28508-28508/? D/cr_GCore: [LifecycleHook.java:73] onApplicationStateChange
> 10-13 09:11:51.723 28508-28508/? D/cr_GCore: [LifecycleHook.java:79] Application visibilty changed to false. Updating state of 1 client(s).
> 10-13 09:11:51.742 28508-28508/? D/cr_GCore: [LifecycleHook.java:73] onApplicationStateChange
> 10-13 09:11:51.886 871-891/? I/WindowManager: Destroying surface Surface(name=SurfaceView - com.android.chrome/org.chromium.chrome.browser.customtabs.CustomTabActivity) called by com.android.server.wm.WindowStateAnimator.destroySurface:2014 com.android.server.wm.WindowStateAnimator.destroySurfaceLocked:881 com.android.server.wm.WindowState.removeLocked:1449 com.android.server.wm.WindowManagerService.removeWindowInnerLocked:2478 com.android.server.wm.AppWindowToken.destroySurfaces:365 com.android.server.wm.WindowStateAnimator.finishExit:565 com.android.server.wm.WindowStateAnimator.finishExit:516 com.android.server.wm.WindowStateAnimator.stepAnimationLocked:491 
> 10-13 09:11:51.898 871-891/? I/WindowManager: Destroying surface Surface(name=com.android.chrome/org.chromium.chrome.browser.customtabs.CustomTabActivity) called by com.android.server.wm.WindowStateAnimator.destroySurface:2014 com.android.server.wm.WindowStateAnimator.destroySurfaceLocked:881 com.android.server.wm.WindowState.destroyOrSaveSurface:2073 com.android.server.wm.AppWindowToken.destroySurfaces:363 com.android.server.wm.WindowStateAnimator.finishExit:565 com.android.server.wm.WindowStateAnimator.stepAnimationLocked:491 com.android.server.wm.WindowAnimator.updateWindowsLocked:303 com.android.server.wm.WindowAnimator.animateLocked:704 
> 10-13 09:11:51.901 401-401/? W/SurfaceFlinger: couldn't log to binary event log: overflow.
> 10-13 09:11:56.812 28508-28508/? D/cr_GCore: [GoogleApiClientHelper.java:185] Disconnect delay expired.
> ```
>
